### PR TITLE
Support cases where func/class is not in the freevars

### LINKED ISSUE #31: Patching an __init__ that calls super() doesn't work
It fails with

```
  File "/opt/yplan/venv3/lib/python3.5/site-packages/patchy/api.py", line 174, in _set_source
    real_func.__code__ = new_code
ValueError: __init__() requires a code object with 1 free vars, not 0
```

From reading here:
http://stackoverflow.com/questions/4885168/how-to-make-super-work-by-manually-filling-the-class-cell

My general understanding is that to make `super()` work on Py3 they turned all `__init__`'s into closures which source `__class__` from the containing scope (regardless of whether the `__init__` actually uses `super`).
Patchy doesn't currently support closures (in general how would you access them to patch them) and so doesn't preserve the `co_freevars` on the code object.

### PR DESCRIPTION
**Pull request by graingert**
Fixes #31.

### CODE REVIEW

File: patchy/api.py [hunk 1]
```diff
@@ -273,7 +285,14 @@ def _process_function():
...
 
     six.exec_(new_code, func.__globals__, localz)
-    new_func = localz['__patchy_freevars__']()
+    to_restore_n = class_name or func.__name__
+    to_restore_v = func.__globals__.get(to_restore_n)
+
+    try:
+        new_func = localz['__patchy_freevars__']()
+    finally:
+        if to_restore_v:
```
[the remark this answers is not in the document]
Comment (graingert) [reply patchy/api.py]: chainmap doesn't work - exec wants a real dict

**Comment by graingert** [comment]
@adamchainz Ok I've finished making changes on this

**Comment by graingert** [comment]
@adamchainz Ok I've finished making changes on this

File: tests/test_patch.py [hunk 2]
```diff
@@ -222,10 +221,45 @@ def __init__(self):
...
+    patchy.patch(factorial, """\
+        @@ -2,4 +2,4 @@
+        -    if n == 1:
+        -        return 1
+        -    else:
+        -        return n * factorial(n-1)
+        +   if n <= 1:
+        +       return n
+        +   else:
+        +       return(factorial(n-1) + factorial(n-2))
```
Comment (adamchainz) [inline tests/test_patch.py]: remove those weird brackets

File: tests/test_patch.py [hunk 3]
```diff
@@ -222,10 +221,45 @@ def __init__(self):
...
     a = Artist()
+    assert mod.Artist == Artist
     assert a.base_prop == 'yo'
     assert a.prop == 'new'
 
 
+def test_patch_recursive_module_level(tmpdir):
+    """
+    Module level recursive fns do not have a freevar for their name, whilst
+    fns defined in a scope do...
```
Comment (adamchainz) [inline tests/test_patch.py]: fn -> function

File: patchy/api.py [hunk 4]
```diff
@@ -257,8 +264,13 @@ def patched_func(self):
...
     def _process_function():
         _def, _ast, fv_body = _process_freevars()
-        ret = '    return {name}'.format(name=func.__name__)
-        to_parse = '\n'.join([_def] + fv_body + ['    pass', ret])
+        name = func.__name__
+        ret = '    return {name}'.format(name=name)
+        _global = (
+            [] if name in func.__code__.co_freevars
+            else ['    global {name}'.format(name=name)]
+        )
```
Comment (adamchainz) [inline patchy/api.py]: unternary this to

```python
if name in func.__code__.co_freevars:
    _global = []
else:
    _global = ['    global {}'.format(name)]
```

File: patchy/api.py [hunk 5]
```diff
@@ -245,7 +245,14 @@ def patched_func(self):
         """
 
         _def, _ast, fv_body = _process_freevars()
-        class_src = '    class {name}(object):\n        pass'.format(name=class_name)
+        _global = (
+            '' if class_name in func.__code__.co_freevars
+            else '    global {name}\n'.format(name=class_name)
+        )
```
Comment (adamchainz) [inline patchy/api.py]: unternary this to:

```python
if class_name in func.__code__.co_freevars:
    _global = ''
else:
    _global = '    global {}\n'.format(name)
```

File: tests/test_patch.py (hunk 3 above)
Comment (graingert) [reply tests/test_patch.py]: done

**Comment by graingert** [comment]
done

File: HISTORY.rst [hunk 6]
```diff
@@ -8,6 +8,9 @@ Pending Release
 
 .. Insert new release notes below this line
 
+* Support cases where the function or class name is not in the freevars. This
+  allows patching `__init__` of a class, or patching a recursive module-level function.
```
Comment (adamchainz) [inline HISTORY.rst]: double backticks for code objects in rst

Comment (graingert) [reply HISTORY.rst]: done

**Comment by adamchainz** [comment]
This and #153 released in version 1.5.0, thanks!

**Pull request merged**